File: supply-chain/java/src/main/java/com/arcadedb/examples/SupplyChain.java
package com.arcadedb.examples;

import com.arcadedb.query.sql.executor.Result;
import com.arcadedb.query.sql.executor.ResultSet;
import com.arcadedb.remote.RemoteDatabase;

public class SupplyChain {

  private static final String HOST     = System.getenv().getOrDefault("ARCADEDB_HOST", "localhost");
  private static final int    PORT     = Integer.parseInt(System.getenv().getOrDefault("ARCADEDB_PORT", "2480"));
  private static final String DB_NAME  = "SupplyChain";
  private static final String USER     = System.getenv().getOrDefault("ARCADEDB_USER", "root");
  private static final String PASSWORD = System.getenv().getOrDefault("ARCADEDB_PASS", "arcadedb");

  public static void main(String[] args) {
    try (RemoteDatabase db = new RemoteDatabase(HOST, PORT, DB_NAME, USER, PASSWORD)) {
      tryRun(() -> runQuery1MultiTierDiscovery(db), "Query 1");
      tryRun(() -> runQuery2BlastRadius(db), "Query 2");
      tryRun(() -> runQuery3DisruptionDetection(db), "Query 3");
      tryRun(() -> runQuery4AlternativeSourcing(db), "Query 4");
      tryRun(() -> runQuery5Traceability(db), "Query 5");
    }
    System.out.println("\nAll queries complete.");
  }

  private static void tryRun(Runnable r, String name) {
    try {
      r.run();
    } catch (Exception e) {
      System.err.println("[" + name + " FAILED] " + e.getMessage());
    }
  }

  // Query 1: Multi-Tier Supplier Discovery
  private static void runQuery1MultiTierDiscovery(RemoteDatabase db) {
    printHeader("Query 1: Multi-Tier Supplier Discovery",
        "Find all suppliers (up to 4 tiers) feeding into Widget Pro X.");

    String cypher =
        """
            MATCH (p:Product {sku: 'WIDGET-PRO-X'})
                  <-[:CONTAINS]-(c:Component)
                  <-[:SUPPLIES*1..4]-(s:Supplier)
            RETURN DISTINCT s.name, s.country, s.risk_score
            ORDER BY s.risk_score DESC""";

    try (ResultSet rs = db.query("cypher", cypher)) {
      while (rs.hasNext()) {
        Result r = rs.next();
        System.out.printf("  %-25s | %-15s | risk: %s%n",
            r.getProperty("s.name"),
            r.getProperty("s.country"),
            r.getProperty("s.risk_score"));
      }
    }
  }

  // Query 2: Blast Radius Analysis
  private static void runQuery2BlastRadius(RemoteDatabase db) {
    printHeader("Query 2: Blast Radius Analysis",
        "If Shenzhen Micro Ltd is disrupted, which products are affected?");

    String cypher =
        """
            MATCH (s:Supplier {name: 'Shenzhen Micro Ltd'})
                  -[:SUPPLIES]->(c:Component)
                  -[:CONTAINS]->(p:Product)
            OPTIONAL MATCH (c)<-[:ALTERNATIVE_FOR]-(alt:Supplier)
            RETURN c.name AS component, p.name AS product, collect(alt.name) AS alternatives""";

    try (ResultSet rs = db.query("cypher", cypher)) {
      while (rs.hasNext()) {
        Result r = rs.next();
        System.out.printf("  %-20s | %-20s | alternatives: %s%n",
            r.getProperty("component"),
            r.getProperty("product"),
            r.getProperty("alternatives"));
      }
    }
  }

  // Query 3: Delivery Disruption Detection
  private static void runQuery3DisruptionDetection(RemoteDatabase db) {
    printHeader("Query 3: Delivery Disruption Detection",
        "Identify suppliers with delivery issues from DeliveryMetric records.");

    String sql =
        """
            SELECT supplierId,
                   avg(lead_time_hrs) AS avg_lead_time,
                   sum(CASE WHEN delayed = true THEN 1 ELSE 0 END) AS total_delayed,
                   count(*) AS total_deliveries
            FROM DeliveryMetric
            GROUP BY supplierId
            ORDER BY total_delayed DESC""";

    try (ResultSet rs = db.query("sql", sql)) {
      while (rs.hasNext()) {
        Result r = rs.next();
        System.out.printf("  %-25s | avg: %6.0f hrs | delayed: %s/%s%n",
            r.getProperty("supplierId"),
            ((Number) r.getProperty("avg_lead_time")).doubleValue(),
            r.getProperty("total_delayed"),
            r.getProperty("total_deliveries"));
      }
    }
  }

  // Query 4: Vector-Based Alternative Sourcing
  private static void runQuery4AlternativeSourcing(RemoteDatabase db) {
    printHeader("Query 4: Vector-Based Alternative Sourcing",
        "Find suppliers with capabilities similar to Shenzhen Micro Ltd [0.9, 0.2, 0.1, 0.1].");

    String sql =
        """
            SELECT name, country, risk_score
            FROM Supplier
            WHERE status = 'active'
            ORDER BY vectorNeighbors('Supplier[capability_vec]', [0.9, 0.2, 0.1, 0.1], 10) DESC
            LIMIT 5""";

    try (ResultSet rs = db.query("sql", sql)) {
      while (rs.hasNext()) {
        Result r = rs.next();
        System.out.printf("  %-25s | %-15s | risk: %.2f%n",
            r.getProperty("name"),
            r.getProperty("country"),
            ((Number) r.getProperty("risk_score")).doubleValue());
      }
    }
  }

  // Query 5: End-to-End Batch Traceability
  private static void runQuery5Traceability(RemoteDatabase db) {
    printHeader("Query 5: End-to-End Batch Traceability",
        "Trace all raw materials in batch BATCH-2026-0218 through the assembly chain.");

    String cypher =
        """
            MATCH (p:Product {batch: 'BATCH-2026-0218'})
                  <-[:ASSEMBLED_FROM*1..8]-(material)
            RETURN material.name, material.origin, material.certification, material.lot""";

    try (ResultSet rs = db.query("cypher", cypher)) {
      while (rs.hasNext()) {
        Result r = rs.next();
        System.out.printf("  %-25s | origin: %-10s | cert: %-10s | lot: %s%n",
            r.getProperty("material.name"),
            r.getProperty("material.origin"),
            r.getProperty("material.certification"),
            r.getProperty("material.lot"));
      }
    }
  }

  private static void printHeader(String title, String description) {
    System.out.println("\n" + "=".repeat(70));
    System.out.println("  " + title);
    System.out.println("  " + description);
    System.out.println("=".repeat(70));
  }
}

File: supply-chain/js/supply-chain.js
const { Client } = require('pg');

const HOST     = process.env.ARCADEDB_HOST    || 'localhost';
const PG_PORT  = process.env.ARCADEDB_PG_PORT || '5432';
const DB_NAME  = 'SupplyChain';
const USER     = process.env.ARCADEDB_USER    || 'root';
const PASSWORD = process.env.ARCADEDB_PASS    || 'arcadedb';

function printHeader(title, description) {
  console.log('\n' + '='.repeat(70));
  console.log('  ' + title);
  console.log('  ' + description);
  console.log('='.repeat(70));
}

async function tryRun(fn, name) {
  try {
    await fn();
  } catch (e) {
    console.error('[' + name + ' FAILED] ' + e.message);
  }
}

// Query 1: Multi-Tier Supplier Discovery (SQL MATCH — Postgres protocol only supports SQL)
async function runQuery1(client) {
  printHeader('Query 1: Multi-Tier Supplier Discovery',
    'Find all suppliers (up to 4 tiers) feeding into Widget Pro X.');

  const sql = `
    SELECT DISTINCT s.name, s.country, s.risk_score
    FROM (
      MATCH {type: Product, where: (sku = 'WIDGET-PRO-X')}
            .in('CONTAINS'){as: c}
            .in('SUPPLIES'){as: s, while: ($depth < 4)}
      RETURN s
    )
    ORDER BY s.risk_score DESC`;

  const res = await client.query(sql);
  for (const row of res.rows) {
    console.log('  %-25s | %-15s | risk: %s',
      row['s.name'] || row.name, row['s.country'] || row.country, row['s.risk_score'] || row.risk_score);
  }
}

// Query 2: Blast Radius Analysis (two SQL MATCH queries — no OPTIONAL MATCH in SQL)
async function runQuery2(client) {
  printHeader('Query 2: Blast Radius Analysis',
    'If Shenzhen Micro Ltd is disrupted, which products are affected?');

  const sqlAffected = `
    SELECT c.name AS component, p.name AS product
    FROM (
      MATCH {type: Supplier, where: (name = 'Shenzhen Micro Ltd')}
            .out('SUPPLIES'){as: c}
            .out('CONTAINS'){as: p}
      RETURN c, p
    )`;

  const sqlAlternatives = `
    SELECT alt.name AS alternative, c.name AS component
    FROM (
      MATCH {type: Supplier, where: (name = 'Shenzhen Micro Ltd')}
            .out('SUPPLIES'){as: c}
            .in('ALTERNATIVE_FOR'){as: alt}
      RETURN alt, c
    )`;

  const affected = await client.query(sqlAffected);
  const alternatives = await client.query(sqlAlternatives);

  const altMap = {};
  for (const row of alternatives.rows) {
    const comp = row.component;
    if (!altMap[comp]) altMap[comp] = [];
    altMap[comp].push(row.alternative);
  }

  for (const row of affected.rows) {
    const alts = altMap[row.component] || [];
    console.log('  %-20s | %-20s | alternatives: [%s]',
      row.component, row.product, alts.join(', '));
  }
}

// Query 3: Delivery Disruption Detection (pure SQL — identical to shell/Java)
async function runQuery3(client) {
  printHeader('Query 3: Delivery Disruption Detection',
    'Identify suppliers with delivery issues from DeliveryMetric records.');

  const sql = `
    SELECT supplierId,
           avg(lead_time_hrs) AS avg_lead_time,
           sum(CASE WHEN delayed = true THEN 1 ELSE 0 END) AS total_delayed,
           count(*) AS total_deliveries
    FROM DeliveryMetric
    GROUP BY supplierId
    ORDER BY total_delayed DESC`;

  const res = await client.query(sql);
  for (const row of res.rows) {
    console.log('  %-25s | avg: %6s hrs | delayed: %s/%s',
      row.supplierid || row.supplierId,
      Number(row.avg_lead_time).toFixed(0),
      row.total_delayed,
      row.total_deliveries);
  }
}

// Query 4: Vector-Based Alternative Sourcing (pure SQL — identical to shell/Java)
async function runQuery4(client) {
  printHeader('Query 4: Vector-Based Alternative Sourcing',
    'Find suppliers with capabilities similar to Shenzhen Micro Ltd [0.9, 0.2, 0.1, 0.1].');

  const sql = `
    SELECT name, country, risk_score
    FROM Supplier
    WHERE status = 'active'
    ORDER BY vectorNeighbors('Supplier[capability_vec]', [0.9, 0.2, 0.1, 0.1], 10) DESC
    LIMIT 5`;

  const res = await client.query(sql);
  for (const row of res.rows) {
    console.log('  %-25s | %-15s | risk: %s',
      row.name, row.country, row.risk_score);
  }
}

// Query 5: End-to-End Batch Traceability (SQL MATCH)
async function runQuery5(client) {
  printHeader('Query 5: End-to-End Batch Traceability',
    'Trace all raw materials in batch BATCH-2026-0218 through the assembly chain.');

  const sql = `
    SELECT name, origin, certification, lot
    FROM (
      MATCH {type: Product, where: (batchId = 'BATCH-2026-0218')}
            .in('ASSEMBLED_FROM'){as: material, while: ($depth < 8)}
      RETURN material.name AS name, material.origin AS origin,
             material.certification AS certification, material.lot AS lot
    )`;

  const res = await client.query(sql);
  for (const row of res.rows) {
    console.log('  %-25s | origin: %-10s | cert: %-10s | lot: %s',
      row.name, row.origin || '-', row.certification || '-', row.lot || '-');
  }
}

async function main() {
  const client = new Client({
    host: HOST,
    port: parseInt(PG_PORT),
    database: DB_NAME,
    user: USER,
    password: PASSWORD,
  });

  await client.connect();
  console.log('Connected to ArcadeDB via PostgreSQL protocol on port ' + PG_PORT);

  try {
    await tryRun(() => runQuery1(client), 'Query 1');
    await tryRun(() => runQuery2(client), 'Query 2');
    await tryRun(() => runQuery3(client), 'Query 3');
    await tryRun(() => runQuery4(client), 'Query 4');
    await tryRun(() => runQuery5(client), 'Query 5');
  } finally {
    await client.end();
  }
  console.log('\nAll queries complete.');
}

main();
